pkg/utils/helper.go

```diff
@@ -132,28 +135,57 @@ func GetHostName() (string, error) {
 	return hostName.HostName, nil
 }
 
-// GetAPIServerFQDN gets the API Server FQDN from the kubeconfig file
-func GetAPIServerFQDN() (string, error) {
-	output, err := RunCommandOnHost("cat", "/var/lib/kubelet/kubeconfig")
-
-	if err != nil {
-		return "", fmt.Errorf("Can't open kubeconfig file: %+v", err)
-	}
-
+//ParseAPIServerFQDNFromKubeConfig parses a kubeConfig and returns the APIServerFQDN
+func ParseAPIServerFQDNFromKubeConfig(output string) (string, error) {
 	lines := strings.Split(output, "\n")
 	for _, line := range lines {
 		index := strings.Index(line, "server: ")
 		if index >= 0 {
 			fqdn := line[index+len("server: "):]
-			fqdn = strings.Replace(fqdn, "https://", "", -1)
-			fqdn = strings.Replace(fqdn, ":443", "", -1)
-			return fqdn, nil
+			fqdnurl, err := url.Parse(fqdn)
+			if err != nil {
+				return "", fmt.Errorf("parse url from fqdn: %s", fmt.Sprint(err)+": "+fqdn)
+			}
+
+			host, _, err := net.SplitHostPort(fqdnurl.Host)
+			if err != nil {
+				return "", fmt.Errorf("split host port from fqdnurl: %s: %w", fqdnurl, err)
+			}
+
+			return host, nil
 		}
 	}
-
 	return "", errors.New("Could not find server definitions in kubeconfig")
 }
 
+//ReadKubeletConfig reads the kubeletConfig from the node
+func ReadKubeletConfig() (string, error) {
+	var result error
+	output, err := RunCommandOnHost("cat", "/var/lib/kubelet/kubeconfig")
+	if err != nil {
+		result = multierror.Append(result, err)
+		output, err = RunCommandOnHost("cat", "/etc/kubernetes/kubelet.conf")
+		if err != nil {
+			result = multierror.Append(result, err)
+			return "", fmt.Errorf("open kubeconfig file at /etc/kubernetes/kubelet.conf or /var/lib/kubelet/kubeconfig\": %+v", result)
+		}
+	}
+	return output, nil
+}
+
+// GetAPIServerFQDN gets the API Server FQDN from the kubeconfig file
+func GetAPIServerFQDN() (string, error) {
+	output, err := ReadKubeletConfig()
+	if err != nil {
+		return "", err
+	}
+	fqdn, err := ParseAPIServerFQDNFromKubeConfig(output)
```

Comment:
agree - this should be private. Moving helper_test to same folder so it can still call it

Reply:
💡 I see that basically the whole location of test files are moved from location to just test `private-method` instead like mentioned above test the `GetAPIServerFQDN()`, keep the test suite location as `tests` folder where all the tests will reside, but test the `public-scope methods`.

Reply:
Its pretty standard golang project structure to put unit test files in the same folder the files they are testing. All AKS projects I have seen are like this. Any particular reason you want to keep tests in a separate folder?

I'm not convinced there is never a case for unit testing private (internal) methods. In this case I think it is the appropriate choice - otherwise we will have to add significant code to mock out and or/ redesign the internal representation just to maintain the current level of testing and with no functional improvements. The language explicitly supports this capability so we should use it where appropriate.

Furthermore this code will be significantly refactored very soon when we change it to use a volume mount rather than directly exec'ing commands on the host. We shouldn't overinvest in its current form unless the benefits are obvious. What are the benefits of only testing the public API here?

Reply:
💡 I think this might help in [better summarizing the point of view which I am coming from](https://softwareengineering.stackexchange.com/questions/100959/how-do-you-unit-test-private-methods). Unless your private method is so complex that it needs its own unit-test. WE can totally rationalize that in here and move forward. What do you think?

>>You generally don't unit test private methods directly. Since they are private, consider them an implementation detail. Nobody is ever going to call one of them and expect it to work a particular way.

>> You should instead test your public interface. If the methods that call your private methods are working as you expect, you then assume by extension that your private methods are working correctly.

Reply:
I think this is complicated enough that  it's worth unit testing the internal method in isolation. If we didn't know it was going to be refactored soon then I would consider pulling it out into a separate package or something, but don't want to over invest.

Also note the closest parallel in Java is "no access modifier defined" and for C# is "internal" keyword. Not "private".

So are we good for now to leave this as is?

Reply:
Sounds good 👍